File: haskell/heuler/src/Euler/Pr007.hs
----------------------------------------------------------------------
-- By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13, we
-- can see that the 6^(th) prime is 13.
--
-- What is the 10001^(st) prime number?

module Euler.Pr007 (pr007) where

import Euler.Problem

pr007 :: Problem
pr007 = Problem { prNumber = 7, prOp = return answer, prCorrect = 104743 }

answer :: Integer
answer = primes !! 10000

-- Classic prime generator.  Not very efficient as the numbers grow
-- larger.
primes :: [Integer]
primes = 2 : 3 : [x | x <- [5..], isPrime x]

isPrime :: Integer -> Bool
isPrime n = filter (\x -> n `mod` x == 0) factors == []
   where
      factors = takeWhile (<= (floor $ sqrt $ fromInteger n)) primes

File: swift/Sources/pr007.swift
// Problem 7
//
// 28 December 2001
//
// By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13, we can see
// that the 6th prime is 13.
//
// What is the 10 001st prime number?
//
// 104743

struct Pr007: Problem {
    typealias T = Int
    let number = 7
    let expected = 104743

    func run() -> Int {
        let sieve = Sieve()
        var prime = 2
        var count = 1

        while count < 10001 {
            prime = sieve.nextPrime(prime)
            count += 1
        }

        return prime
    }
}
